addons/controllers/packageinstallstatus_controller.go

```diff
@@ -0,0 +1,442 @@
+	// is installed in the workload cluster's namespace in the management cluster.
+	// For all other packages belonging to the workload cluster, remoteClient should be used for fetching packageMetadata
+	if clusterRole == clusterRoleManagement || strings.Contains(pkgName, kappCtrlPkgPrefix) {
+		Client = r.aggregatedAPIResourcesClient
+	}
+
+	// Note: pkgi name is determined from pkg.RefName. That's why we rely on the existence of package metadata here
+	pkgRefName, pkgVersion, err := util.GetPackageMetadata(ctx, Client, pkgName, pkgiNamespace)
+	if pkgRefName == "" || pkgVersion == "" || err != nil {
+		log.Error(err, fmt.Sprintf("unable to fetch Package.Spec.RefName or Package.Spec.Version from Package '%s/%s'", pkgiNamespace, pkgName))
+		return false, err
+	}
+	pkgShortname = strings.Split(pkgRefName, ".")[0]
+
+	// package install name for core/additional packages in the workload cluster follows the <cluster name>-<addon short name> naming convention
+	// for management cluster, the naming convention for core/additional package installs is just <addon short name>
+	if clusterRole == clusterRoleWorkload {
+		pkgiName = util.GeneratePackageInstallName(clusterName, pkgRefName)
+	} else if clusterRole == clusterRoleManagement {
+		pkgiName = pkgShortname
+	}
+
+	pkgi := &kapppkgiv1alpha1.PackageInstall{}
+	objectKey := client.ObjectKey{Namespace: pkgiNamespace, Name: pkgiName}
+
+	// Note: use of the client (remote or not) depends on whether the corresponding cluster is
+	// workload or management which is determined in the Reconcile function
+	if err := c.Get(ctx, objectKey, pkgi); err != nil {
+		log.Error(err, fmt.Sprintf("unable to get PackageInstall '%s/%s'", pkgiNamespace, pkgiName))
+		return false, err
+	}
+
+	// For each package, create a single summary condition from the condition slice
+	pkgiCondition := util.SummarizeAppConditions(pkgi.Status.Conditions)
+
+	// Note: in case of encountering an unknown PackageInstall condition, just return err=nil and proceed with handling the next package
+	if pkgiCondition.Type == util.UnknownCondition {
+		log.Error(fmt.Errorf("unknown condition type for '%s/%s'", pkgi.Name, pkgi.Name), "unknown condition")
+		return false, nil
+	}
+
+	condition := runtanzuv1alpha3.Condition{
+		Type: runtanzuv1alpha3.ConditionType(strings.Title(pkgShortname)) + "-" +
+			runtanzuv1alpha3.ConditionType(pkgiCondition.Type),
+		Status:             pkgiCondition.Status,
+		UsefulErrorMessage: util.GetKappUsefulErrorMessage(pkgi.Status.UsefulErrorMessage),
+		LastTransitionTime: metav1.NewTime(time.Now().UTC().Truncate(time.Second)),
+	}
+
+	// Only add a new condition entry for the PackageInstall in the clusterBootstrapStatus in case it doesn't already exist
+	// If it does, just update it with the new condition
+	var conditionExists bool
+	var conditionChanged bool
+	for i, existingCond := range clusterBootstrapStatus.Conditions {
```

Comment:
The logic here seems a bit complex to me. Maybe introducing a map would help here.